src/cuda/api/array.hpp

```diff
@@ -0,0 +1,153 @@
+/**
+ * @file array.hpp
+ * 
+ * @brief A CUDA array wrapper class. Especially useful
+ * if used in combination with CUDA textures.
+ *
+ */
+
+#pragma once
+#ifndef CUDA_API_WRAPPERS_ARRAY_HPP
+#define CUDA_API_WRAPPERS_ARRAY_HPP
+
+#include <cuda/api/current_device.hpp>
+#include <cuda/api/error.hpp>
+
+#include <cuda_runtime.h>
+#include <cuda_runtime_api.h>
+
+#include <algorithm>
+#include <array>
+#include <numeric>
+
+namespace cuda {
+
+template<bool AssumedCurrent> class device_t;
+
+namespace array {
+
+namespace detail {
+
+/**
+ * @brief Do not use directly
+ *
+ */
+template <typename T, size_t NDIMS>
+class array_base {
+
+protected:
+	array_base(dimensions_t<NDIMS> dims, cudaArray* array_ptr) :
+	    dims_(dims),
+	    array_ptr_(array_ptr) {}
+
+public:
+	array_base()                        = delete;
+	array_base(const array_base& other) = delete;
+
+	array_base(array_base&& other) :
+	    array_ptr_(other.array_ptr_),
+	    dims_(other.dims_) {
+		other.array_ptr_ = nullptr;
+		other.dims_      = {0, 0, 0};
+	}
+
+	~array_base() {
+		if (array_ptr_) {
+			auto status = cudaFreeArray(array_ptr_);
+			throw_if_error(status, "failed freeing CUDA array");
+		}
+	}
+
+	cudaArray* get() const noexcept { return array_ptr_; }
+
+	dimensions_t<NDIMS> dims() const noexcept { return dims_; }
+
+	size_t size() const noexcept {
+		using dimensions_value_type = typename dimensions_t<NDIMS>::value_type;
+		return std::accumulate(
+		    dims_.begin(), dims_.end(), static_cast<dimensions_value_type>(1),
+		    [](const dimensions_value_type& a, const dimensions_value_type& b) { return a * b; });
+	}
+
+	size_t size_bytes() const noexcept { return size() * sizeof(T); }
+
+protected:
+	dimensions_t<NDIMS> dims_;
+	cudaArray*          array_ptr_;
+};
+
+} // namespace detail
+
+/**
+ * @brief Wrapper for 2D and 3D arrays
+ * 
+ * Useful in combination with texture memory.
+ * 
+ * @note Only template specializations of this class can be created, because
+ * CUDA does not support any number of dimensions for a CUDA array.
+ */
+template <typename T, size_t NDIMS>
+class array_t {
+public:
+	array_t() = delete;
+};
+
+template <typename T>
+class array_t<T, 3> : public detail::array_base<T, 3> {
+
+private:
+
+	template<bool AssumedCurrent>
+	static cudaArray*
+	malloc_3d_cuda_array(const device_t<AssumedCurrent>& device, dimensions_t<3> dims) {
```

Comment:
- If you do not like the name of the function: how should I name it?
- It returns a raw cuda array pointer, which is not needed by any other internal function of this library, nor should any user of the library use this function directly. Thus I think `array.hpp` is a reasonable location for this function.

Reply:
We have `allocate()` methods and functions (e.g. in `memory.hpp`) - consider using that. Also, consider using namespaces instead of some parts of the name.

About the location - I guess it's debatabe. On the one hand, it's an allocation of memory; and on the other hand it's only used in arrays. What about the namespace, then? `cuda::memory::allocate_array<3>()` ? Or do you think `cuda::array::allocate<3>`? Also, should we perhaps have an enum under `cuda::array` so that we can use `cuda::array::3D` and `cuda::array::2D` ? Yet another option is to have the allocation named `allocate()` with no mention of the dimensions, and have an overload set with one function for 2D taking 2 dimension parameters and one for 3D taking 3 of them. And yet another option is to have a single function, but take `dimensions_t` and a dimensionality to use (which would be either 2 or 3).

Reply:
So, never mind the location right now. Let is stay here for the time being and we can think about this sometime later.